README.md
# Maxfucktor-hs
## What
A dead simple BF compiler written this time in Haskell.
## How
In the project's root you will find a `build.sh` script, that being invoked first builds the compiler itself and then immediatelly shoves provided BF source into it (requires `stack` and `yasm`).
```bash
>> ./build.sh bf_src/helloworld.b
>> ./helloworld.b.exe 
Hello World!
>>
```
Resulting ASM code should look like this:
```nasm
global _start
section .data
  memory: times 32768 db 0
section .text
_start:
  mov rsi, memory
  mov rdx, 1 ;; rdx wont change during the runtime
  mov rdi, 1 ;; rdi represents an io descriptor, typically 1 or 0
  jmp run
exit:
  mov rax, 60
  mov rdi, 0
  syscall
run:
add byte [rsi], byte 10
l1:
cmp byte [rsi], byte 0
je l2
add rsi, 1
add byte [rsi], byte 7
add rsi, 1
add byte [rsi], byte 10
add rsi, 1
add byte [rsi], byte 3
add rsi, 1
add byte [rsi], byte 1
sub rsi, 4
sub byte [rsi], byte 1
jmp l1
l2:
add rsi, 1
add byte [rsi], byte 2
mov rax, 1
syscall
add rsi, 1
add byte [rsi], byte 1
mov rax, 1
syscall
add byte [rsi], byte 7
mov rax, 1
syscall
mov rax, 1
syscall
add byte [rsi], byte 3
mov rax, 1
syscall
add rsi, 1
add byte [rsi], byte 2
mov rax, 1
syscall
sub rsi, 2
add byte [rsi], byte 15
mov rax, 1
syscall
add rsi, 1
mov rax, 1
syscall
add byte [rsi], byte 3
mov rax, 1
syscall
sub byte [rsi], byte 6
mov rax, 1
syscall
sub byte [rsi], byte 8
mov rax, 1
syscall
add rsi, 1
add byte [rsi], byte 1
mov rax, 1
syscall
jmp exit
```
Not great, not terrible, I know I know.


ChangeLog.md
# Changelog for hask

## Unreleased changes


src/Maxfucktor/Parser.hs
{-# LANGUAGE TupleSections #-}


module Maxfucktor.Parser (AST(..), wholeProgram, Parser(runParser)) where


import Control.Applicative ( Alternative(some, many, (<|>)) )
import Data.Maybe ( isJust )


data AST = 
    Inc     Int
  | Dec     Int
  | GoLeft  Int
  | GoRight Int
  | Input   Int
  | Output  Int
  | Loop    [AST]
  deriving (Show, Eq)


-- i : input type
-- o : output type
newtype Parser i o = Parser {runParser :: [i] -> (Maybe o, [i])}


instance Functor (Parser input) where
  fmap f parser = 
    Parser $ \input ->
      let (result, input_rest) = runParser parser input in
        (f <$> result, input_rest)


instance Applicative (Parser input) where
  pure a = Parser (Just a,)
  pf <*> pa = 
    Parser $ \input ->
      let (f, input') = runParser pf input in
        let (a, input'') = runParser pa input' in
          (
            do
              func <- f
              value <- a
              Just $ func value,
            input''
          )


punit :: Parser Char ()
punit = return ()


instance Monad (Parser input) where
  pa >>= f =
    Parser $ \input ->
      let (a, input') = runParser pa input in
        case a of
          Nothing -> (Nothing, input')
          Just value -> runParser (f value) input'


instance Alternative (Parser input) where
  p0 <|> p1 =
    Parser $ \input ->
      let res@(r0, input') = runParser p0 input in
        case r0 of
          Just value -> res
          Nothing -> runParser p1 input


predicate :: (a -> Bool) -> Parser a a
predicate f = 
  Parser $ \input ->
    let fail = (Nothing, input) in
    if null input
      then fail
      else
        if f $ head input
          then (Just $ head input, tail input)
          else fail


-- BF loop [%code-here%]
loopBody :: Parser Char AST
loopBody = do
  predicate (== '[')
  matched_code <- many code
  predicate (== ']')
  return $ Loop matched_code


-- any BF code
code :: Parser Char AST
code =
  foldl (<|>) loopBody $
  map 
    (\(sym, ast) -> ast . length <$> some (predicate (== sym)))
    [
      ('+', Inc),
      ('-', Dec),
      ('>', GoRight),
      ('<', GoLeft),
      ('.', Output),
      (',', Input)
    ]


noLeftOvers :: Eq b => Parser a b -> Parser a b
noLeftOvers p=
  Parser $ \input ->
    let success@(inner_result, leftover) = runParser p input in
      let fail = (Nothing, leftover) in
      if isJust inner_result && null leftover
        then success
        else fail


wholeProgram :: Parser Char [AST]
wholeProgram = noLeftOvers $ many code


src/Maxfucktor/Generator.hs
{-# OPTIONS_GHC -fwarn-incomplete-patterns #-}
module Maxfucktor.Generator (renderProgram) where


import qualified Maxfucktor.Optimizer as Opt
import Control.Monad.State (State, get, put, evalState)


type Strings = [[Char]]


data Op = Add | Sub deriving (Eq, Show)


-- This type is used to enumerate lables in resulting ASM code
newtype BlockId = BlockId Int deriving Eq


-- It might be a good idea to implement the "Enum BlockId" instead of this
nextContid :: BlockId -> BlockId
nextContid (BlockId value) = BlockId $ value + 1


instance Show BlockId where
    show (BlockId value) = "l" ++ show value 


-- Converts a single "optimized" AST node into ASM code building action
renderNode :: Opt.AST -> State BlockId Strings
renderNode node =
    case node of
        Opt.Inc rep ->
            return ["add byte [rsi], byte " ++ show rep]
        Opt.Dec rep ->
            return ["sub byte [rsi], byte " ++ show rep]
        Opt.GoRight rep ->
            return ["add rsi, " ++ show rep]
        Opt.GoLeft  rep ->
            return ["sub rsi, " ++ show rep]
        Opt.Output  rep ->
            return $
            concat
                [
                    [ "mov rax, 1", "syscall" ] | counter <- [1..rep]
                ]
        Opt.Input  rep ->
            return $
            concat [
                [
                    "mov rax, 0",
                    "mov rdi, 0",
                    "syscall",
                    "mov rdi, 1"
                ] | counter <- [1..rep]
            ]
        Opt.Mul shift0 shift1 mul0 mul1 ->
            return $
            let header = [";;; Starting Mul block" ] in
            let footer = [";;; Ending Mul block" ] in
                header ++
                toASMBlock mul0 shift0 ++
                toASMBlock mul1 shift1 ++
                ["mov [rsi], byte 0"] ++
                footer
            where
                toASMBlock mul shift = 
                    ["movzx rax, byte [rsi]"] ++
                    (
                        if mul > 1
                        then [
                            "movzx r12, byte " ++ show mul,
                            "mul r12b"
                        ]
                        else mempty 
                    ) ++
                    [
                        "add byte [rsi+" ++ show shift ++"], al"
                    ]
        Opt.Add value ->
            do
                -- call the 'next' function
                current_id <- get
                put $ nextContid current_id
                -- end of 'next' invocation
                return $ renderSubAdd value current_id Add
        Opt.Sub value ->
            do
                -- call the 'next' function
                current_id <- get
                put $ nextContid current_id
                -- end of 'next' invocation
                return $ renderSubAdd value current_id Sub
        Opt.Drop ->
            return
            [
                "mov byte [rsi], byte 0"
            ]
        Opt.Loop nodes ->
            do
                -- call the 'next' function twice
                this_id <- get
                put $ nextContid this_id
                cont_id <- get
                put $ nextContid cont_id
                -- end of 'next' invocations
                innerNodesCode <- mapM renderNode nodes
                return $
                  initLoop this_id cont_id ++ -- loop initialization code
                  concat innerNodesCode    ++ -- code for the loop body itself
                  loopBack this_id cont_id    -- loop back code
            where
                initLoop this_id cont_id = [
                    show this_id ++ ":",
                    "cmp byte [rsi], byte 0",
                    "je " ++ show cont_id
                    ]
                loopBack this_id cont_id = [
                    "jmp " ++ show this_id,
                    show cont_id ++ ":"
                    ]
                jump target_id = ["jmp " ++ show target_id ]
    where
        renderSubAdd :: Int -> BlockId -> Op -> Strings
        renderSubAdd value cont_id op =
            let opRepr = if op == Add then "add" else "sub" in
            let header = [";;; Starting Add block"] in
            let footer = [";;; Ending Add block"] in
                header ++
                [
                    "cmp byte [rsi], byte 0",
                    "je .skip" ++ show cont_id,
                    "movzx r11, byte [rsi]",
                    "mov byte [rsi], byte 0",
                    opRepr ++ " byte [rsi+(" ++ show value ++ ")], r11b",
                    ".skip" ++ show cont_id
                ] ++
                footer


renderProgram :: [Opt.AST] -> Strings
renderProgram nodes = 
    let code = concat $ evalState (mapM renderNode nodes) (BlockId 1) in
        code ++ ["jmp exit"]


src/Maxfucktor/Optimizer.hs
{-# OPTIONS_GHC -fwarn-incomplete-patterns #-}


module Maxfucktor.Optimizer where


import qualified Maxfucktor.Parser as P


data AST = 
    Inc     Int
  | Dec     Int
  | GoLeft  Int
  | GoRight Int
  | Input   Int
  | Output  Int
  | Loop    [AST]
  | Drop
  | Add     Int
  | Sub     Int
  | Mul     {shift0::Int, shift1::Int, mul0::Int, mul1::Int}
  deriving (Show, Eq)


optimize :: P.AST -> AST
optimize astnode = 
    case astnode of
        P.Loop [P.Dec 1, P.GoRight s0, P.Inc 1, P.GoLeft s1] | s0 == s1 -> Add s0
        P.Loop [P.Dec 1, P.GoLeft s0, P.Inc 1, P.GoRight s1] | s0 == s1 -> Add (-s0)
        P.Loop [P.Dec 1, P.GoRight s0, P.Dec 1, P.GoLeft s1] | s0 == s1 -> Sub s0
        P.Loop [P.Dec 1, P.GoLeft s0, P.Dec 1, P.GoRight s1] | s0 == s1 -> Sub (-s0)
        P.Loop
            [P.Dec 1, P.GoRight p0, P.Inc p1, P.GoRight p2, P.Inc p3, P.GoLeft p4]
            | p0 + p2 == p4 ->
            Mul p0 (p0+p2) p1 p3
        P.Loop [P.Dec value] | value > 0 -> Drop
        P.Loop contents -> Loop $ map optimize contents
        P.Inc     value -> Inc value
        P.Dec     value -> Dec value
        P.GoLeft  value -> GoLeft value
        P.GoRight value -> GoRight value
        P.Input   value -> Input value
        P.Output  value -> Output value


app/Main.hs
module Main where

import Data.List (intercalate)
import Maxfucktor.Generator (renderProgram)
import Maxfucktor.Optimizer (optimize)
import Maxfucktor.Parser (Parser (runParser), wholeProgram)
import System.Environment (getArgs)


programHeader :: [[Char]]
programHeader = [
  "global _start",
  "section .data",
  "  memory: times 32768 db 0",
  "section .text",
  "_start:",
  "  mov rsi, memory",
  "  mov rdx, 1 ;; rdx wont change during the runtime",
  "  mov rdi, 1 ;; rdi represents an io descriptor, typically 1 or 0",
  "  jmp run",
  "exit:",
  "  mov rax, 60",
  "  mov rdi, 0",
  "  syscall",
  "run:\n"
  ]


main :: IO ()
main =
  do
    args <- getArgs
    datum <- readFile $ head args
    let filtered = filter (`elem` "<>.,[]+-") datum in
      case runParser wholeProgram filtered of
        (Nothing, leftover) ->
          print $ "NoParse error occured, unable to process: " ++ leftover
        (Just ast, _) ->
          let opt_ast = map optimize ast
            in putStr $
            intercalate "\n" programHeader ++
            (intercalate "\n" (renderProgram opt_ast) ++ "\n")
